shared/src/price_estimation/gas.rs

```diff
@@ -0,0 +1,21 @@
+///! constants to estimate gas use in GPv2
+
+/// minimum gas every settlement takes
+pub const SETTLEMENT: u64 =
+    // initial tx gas
+    32_000 +
+    // isSolver
+    7365;
+
+/// gas per trade excluding c20 transfer
+pub const TRADE: u64 =
+    // computeTradeExecutions
+    35_000 +
+    // transferFromAccounts and transferToAccount overhead
+    2 * 3000 +
+    // overhead of one interaction
+    3000;
+
+/// lower bound for an erc20 transfer
+/// TODO: Use an average or median?
+pub const ERC20_TRANSFER: u64 = 20_000;
```

Comment:
@bh2smith Maybe this is something you can help out with?

Reply:
I can put together a query which fetches the the gas cost for transfer of a very very large list of tokens. Will get back to you here with a measure (min, max and median) for us to look at.

Reply:
Thank you. I think we can merge this before but would be nice to have for the future. For a median it might be nice to only pick the first X tokens that make up 90% of trades or something like that.

Reply:
I have put together this query which show the min, max, mean, median and sample size of the gas used for each token on the top 90% of tokens traded on GPv2.

It only samples from transfers as far back as October 15, 2021 at the moment because its quite an intense calculation and takes 4 minutes to run as is.

https://dune.xyz/queries/223162

Of course I will take the median over all at some point, just wanted us to have an idea on a per token basis what we will be working with.

Reply:
This separate query aggregates the data over all tokens.

https://dune.xyz/queries/224033

Reply:
I'm actually thinking we should target the 51 or 52 percentile median so that we are on average estimating slightly above the "break even" point. Like in black Jack.

Reply:
Do these transfers include the 21K gas for a transaction? We probably don't want to include that in this number.

Reply:
Hmm, I think there may be something wrong with the query. Shouldn't the position of the `transfer()` selector be `0` and not strictly greater than `0` (because the first 4 bytes of the call data for transfers should ALWAYS be `0xa9059cbb`):
```
position('\xa9059cbb' in data) = 0
-- Instead of:
position('\xa9059cbb' in data) > 0
```



Also, from looking at the query, it includes the 21K transaction cost that shouldn't be part of this number.

Reply:
I ran the query with the suggestions :point_up: (adjusted `position` condition and 52 percentile median) and got:
```


        maximum |        minimum | median_average |   mean_average |     sample_size
----------------+----------------+----------------+----------------+----------------
       24274287 |          21505 |          46551 | 52542.93944... |         1989470
```

So I would say we can use `25_551`.

Reply:
Very nice!

Reply:
I guess `position() > 0` could also include transactions that had more than just a transfer, but also other stuff. Unfortunately I suspect `position() = 0` will greatly improve this measure, it still might contain transactions that have transfer as the first call. In any case I think we have a reasonable estimate.

Reply:
```suggestion
pub const ERC20_TRANSFER: u64 = 25_551;
```

Reply:
> I guess position() > 0 could also include transactions that had more than just a transfer

I'm not sure I understand... From my understanding of Solidity ABI, all `transfer` transactions are start with calldata `0xa9059cbb` (the "selector"). By "if transaction has more than one call" do you mean something like a multisend? In that case, I don't think these transactions won't be included since the query is specifically for calls to token contracts starting with the transfer selector.

> I guess `position() > 0` could also include transactions that had more than just a transfer, but also other stuff.

The issue for me is that it may include transactions that aren't `transfers` since if an address or and amount includes the bytes `0xa9059cbb` (and 4-byte collision is not that rare) will get included even if they aren't transfers at all.

It could be that I don't understand what the `data` column in Dune means though.

Reply:
The issue is that it seems like `position` uses indices starting at 1. This means that:
```
position("\x0102", "\x03") = 0 and position("\x0102", "\x01") = 1
```

So the query should have used `position() = 1`.